### PR TITLE
add clang-16 build

### PR DESCRIPTION
**Pull request by mikea**
We're having upstream compilation problems with clang-16, adding it here so that I can write unit-tests/add fixes.

### CODE REVIEW

**Comment by kentonv** [comment]
We can replace 14 with 16. The idea is to test the oldest and newest versions that we want to support, we assume everything in between will work...

**Comment by kentonv** [comment]
Oh I see you are actually adding two new variants of the bazel build. Do we really need to cover so many different Clang versions?

File: .github/workflows/quick-test.yml [hunk 1]
```diff
@@ -179,19 +179,23 @@ jobs:
...
     steps:
       - uses: actions/checkout@v3
       - uses: bazelbuild/setup-bazelisk@v2
       - name: install dependencies
         run: |
             export DEBIAN_FRONTEND=noninteractive
-            sudo apt-get install -y build-essential git ${{ matrix.compiler }}
+            sudo apt-get install -y build-essential git
+      - name: install clang
+        uses: egor-tensin/setup-clang@v1
```
Comment (kentonv) [inline .github/workflows/quick-test.yml:196]: I'm not super excited about depending on random third-party repos, especially for something we only need temporarily (until a clang-16 package lands in Ubuntu, presumably).

Comment (kentonv) [reply .github/workflows/quick-test.yml:196]: Actually I see this build is using Ubuntu 20.04, which is pretty old. Did you try 22.04? Maybe it has a clang-16 package already?

Comment (kentonv) [reply .github/workflows/quick-test.yml:196]: OK I see... there is a clang-16 package but only for Ubuntu 23.04 (not LTS) so I guess indeed we'd have to wait another year.

I guess our actions don't have any secrets so there's no security concern here.

Given that we already cover clang 6 and 14 in the other builds, can we have the bazel build focus on 16? I worry we have too many builds...

Comment (mikea) [reply .github/workflows/quick-test.yml:197]: done. Bazel is clang-16 only.

**Pull request merged**